pkg/validators/retry.go

```diff
@@ -0,0 +1,265 @@
+package validators
+
+import (
+	"math/rand"
+	"time"
+
+	"github.com/mt-sre/addon-metadata-operator/pkg/types"
+)
+
+// NewRetryMiddleware applies any given options to an instance of
+// RetryMiddleware and provides defaults for those not provided.
+// A *RetryMiddleware instance is then returned ready to be used
+// as middleware for types.ValidatorRunner's.
+func NewRetryMiddleware(opts ...RetryMiddlewareOption) *RetryMiddleware {
+	var mw RetryMiddleware
+
+	for _, opt := range opts {
+		opt(&mw)
+	}
+
+	if mw.maxRetries == 0 {
+		mw.maxRetries = 5
+	}
+
+	if mw.delayer == nil {
+		mw.delayer = Constant(2 * time.Second)
+	}
+
+	return &mw
+}
+
+// RetryMiddleware implements middleware that conditionally
+// retries types.ValidatorRunner executions based on the result.
+type RetryMiddleware struct {
+	maxRetries int
+	delayer    types.Delayer
+}
+
+func (r *RetryMiddleware) Wrap(v types.ValidatorRunner) types.ValidatorRunner {
+	return types.ValidateFunc(func(mb types.MetaBundle) types.ValidatorResult {
```

Comment:
is extra `types.ValidateFunc` wrapper needed? Compiler will complain if dont respect the `types.ValidatorRunner` interface

Reply:
See comment on `ValidateFunc`